flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/result/CollectStreamResult.java

```diff
@@ -138,7 +146,10 @@ protected boolean isRetrieving() {
 		@Override
 		public void run() {
 			try {
-				deployer.run();
+				// fetch the job execution result, so that an attached cluster will shut down
+				deployer
+						.run()
+						.thenCompose((jobClient -> jobClient.getJobExecutionResult(classLoader)));
```

Comment:
Maybe we should close job client here?

Reply:
You're right. I noticed that working with a `JobClient` that needs to be closed is very cumbersome. The code becomes:
```
deployer
		.deploy()
		.thenCompose((jobClient -> {
			CompletableFuture<JobExecutionResult> result =
					jobClient.getJobExecutionResult(classLoader);
			try {
				jobClient.close();
			} catch (Exception e) {
				throw new CompletionException("Error while closing JobClient.", e);
			}
			return result;
		}));
```
It would be easier to write correct code if the `JobClient` didn't need closing.

Reply:
I give this snippet a think. Here comes some thoughts I'd like to sync our opinion.

1. We don't introduce extra complexity here. Previously we should make sure closing `ClusterClient` properly and right now we close the `JobClient`. It is about lifecycle management that we cannot avoid it conceptually.

2. For code side, we can ease boilerplate code eventually as

```java
.thenCompose((jobClient -> {
    try (jobClient) {
        return jobClient.getJobExecutionResult(classLoader);
    }
}));
```

here we need two changes: (1) Java 9 support resource reference in try-with-resource. Although we cannot bump codebase to Java 11 dropping Java 8 for now, it is a predictable improvement once we move there. (2) `JobClient#close` can be tagged without `throw Exception` and so does `ClusterClient#close`. Their implementations don't throw checked exceptions and we cannot do too much when closing client failed. In this context, we just fail the completable future with the exception as expected.

Reply:
After pushing this change (where I close the `JobClient`) I noticed that this does not actually work. The problem is that if I shut down the `JobClient` then the `thenAccept()` method will not be called because the `ClusterClient` internally shuts down executors and other things, meaning that I never get the result and we just hang.

Reply:
resolved in ec7ba22b2c7a8aa2f9130285f28db1520c2e2442 I think.